src/main/java/io/vertx/core/eventbus/impl/CodecManager.java

```diff
@@ -46,103 +46,111 @@
-    } else if (body instanceof JsonObject) {
-      codec = JSON_OBJECT_MESSAGE_CODEC;
-    } else if (body instanceof JsonArray) {
-      codec = JSON_ARRAY_MESSAGE_CODEC;
-    } else if (body instanceof byte[]) {
-      codec = BYTE_ARRAY_MESSAGE_CODEC;
-    } else if (body instanceof Integer) {
-      codec = INT_MESSAGE_CODEC;
-    } else if (body instanceof Long) {
-      codec = LONG_MESSAGE_CODEC;
-    } else if (body instanceof Float) {
-      codec = FLOAT_MESSAGE_CODEC;
-    } else if (body instanceof Double) {
-      codec = DOUBLE_MESSAGE_CODEC;
-    } else if (body instanceof Boolean) {
-      codec = BOOLEAN_MESSAGE_CODEC;
-    } else if (body instanceof Short) {
-      codec = SHORT_MESSAGE_CODEC;
-    } else if (body instanceof Character) {
-      codec = CHAR_MESSAGE_CODEC;
-    } else if (body instanceof Byte) {
-      codec = BYTE_MESSAGE_CODEC;
-    } else if (body instanceof ReplyException) {
-      codec = defaultCodecMap.get(body.getClass());
-      if (codec == null) {
-        codec = REPLY_EXCEPTION_MESSAGE_CODEC;
-      }
+      return codec;
     } else {
-      codec = defaultCodecMap.get(body.getClass());
-      if (codec == null) {
-        throw new IllegalArgumentException("No message codec for type: " + body.getClass());
+      if (body == null) {
+        return NULL_MESSAGE_CODEC;
+      } else {
+        Class<?> clazz = body.getClass();
+        if (clazz == String.class) {
+          return STRING_MESSAGE_CODEC;
+        } else if (Buffer.class.isAssignableFrom(clazz)) {
+          return BUFFER_MESSAGE_CODEC;
+        } else if (clazz == JsonObject.class) {
+          return JSON_OBJECT_MESSAGE_CODEC;
+        } else if (clazz == JsonArray.class) {
+          return JSON_ARRAY_MESSAGE_CODEC;
+        } else if (clazz == byte[].class) {
+          return BYTE_ARRAY_MESSAGE_CODEC;
+        } else if (clazz == Integer.class) {
+          return INT_MESSAGE_CODEC;
+        } else if (clazz == Long.class) {
+          return LONG_MESSAGE_CODEC;
+        } else if (clazz == Float.class) {
+          return FLOAT_MESSAGE_CODEC;
+        } else if (clazz == Double.class) {
+          return DOUBLE_MESSAGE_CODEC;
+        } else if (clazz == Boolean.class) {
+          return BOOLEAN_MESSAGE_CODEC;
+        } else if (clazz == Short.class) {
+          return SHORT_MESSAGE_CODEC;
+        } else if (clazz == Character.class) {
+          return CHAR_MESSAGE_CODEC;
+        } else if (clazz == Byte.class) {
+          return BYTE_MESSAGE_CODEC;
+        } else if (ReplyException.class.isAssignableFrom(clazz)) {
+          return codecsByClass.getOrDefault(clazz, REPLY_EXCEPTION_MESSAGE_CODEC);
+        } else {
+          MessageCodec codec = resolveCodec(clazz);
+          if (codec == null) {
+            throw new IllegalArgumentException("No message codec for type: " + clazz);
+          }
+          return codec;
+        }
+      }
+    }
+  }
+
+  private MessageCodec resolveCodec(Class<?> clazz) {
```

Comment:
`resolveCodec` will resolve any available default codec, even if it doesn't match the source objects class. As suggested previously, what do you think of adding a:

```java
default boolean supportsGenericTypes {
  return false
}
```

To the `MessageCodec` interface and then changing the implementation of this to:

private MessageCodec resolveCodec(Class<?> clazz) {
 
 
`resolveCodec` will resolve any available default codec, even if it doesn't match the source objects class. As suggested previously, what do you think of adding a:

```java
default boolean supportsGenericTypes {
  return false
}
```

To the `MessageCodec` interface and then changing the implementation of this:

```java
private MessageCodec resolveCodec(Class<?> resolve, Class<?> clazz) {
    MessageCodec codec;
    codec = codecsByClass.get(resolve);
    if (codec != null && !resolve.equals(clazz) && !codec.supportsGenericTypes()) {
       codec = null; // this codec does not support converting generic types
    }
    if (codec == null) {
      Class<?> parent = clazz.getSuperclass();
      if (parent != Object.class && parent != null) {
        codec = resolveCodec(parent, clazz);
      }
    }
    return codec;
  }
```

That'd make it a non-breaking change for one and also make it more safe, because codecs will have to specifically announce if they want to be supporting generic types.